src/rest/connection.cpp

```diff
@@ -0,0 +1,297 @@
+    if (it == this->mDiagInfo->mNodeSet.end())
+        return false;
+    else
+        return true;
+}
+
+void Connection::AddDiag(std::string aRloc, std::string aDiag)
+{
+    this->mDiagInfo->mDiagContent.push_back(aDiag);
+    this->mDiagInfo->mNodeSet.insert(aRloc);
+}
+
+void Connection::SetCallbackFlag(int aFlag)
+{
+    this->mCallbackFlag = aFlag;
+}
+void Connection::SetErrorFlag(int aFlag)
+{
+    this->mEndFlag = aFlag;
+}
+
+void Connection::ResetDiagInfo()
+{
+    this->mDiagInfo.reset();
+    mDiagInfo = std::unique_ptr<DiagInfo>(new DiagInfo(steady_clock::now()));
+}
+
+void Connection::ErrorFormResponse(Response &aResponse)
+{
+    aResponse.SetBody(this->mErrorCode);
+}
+
+void Connection::CallbackFormResponse(Response &aResponse)
+{
+    std::string str = mJsonFormater.VectorToJson(this->mDiagInfo->mDiagContent);
+    aResponse.SetBody(str);
+}
+
+void Connection::SetErrorCode(std::string aErrorCode)
+{
+    this->mErrorCode = aErrorCode;
+}
+
+requestHandler Connection::GetHandler(Request &aRequest)
+{
+    std::string url     = aRequest.getPath();
+    auto        Handler = this->mHander.GetHandler(url);
+    return Handler;
+}
+
+steady_clock::time_point Connection::GetConnectionStartTime()
+{
+    return this->mStartTime;
+}
+
+steady_clock::time_point Connection::GetDiagStartTime()
+{
+    return this->mDiagInfo->mStartTime;
+}
+
+void Connection::SentResponse(Response &aResponse)
+{
+    std::string data = aResponse.SerializeResponse();
+    write(this->mFd, data.c_str(), data.size());
+}
+
+int Connection::OnMessageBegin(http_parser *parser)
+{
+    Request *request = (Request *)parser->data;
+    request->SetComplete(0);
+    return 0;
+}
+
+int Connection::OnStatus(http_parser *parser, const char *at, size_t len)
+{
+    Request *request = (Request *)parser->data;
+    request->SetStatus(at, len);
+    return 0;
+}
+
+int Connection::OnUrl(http_parser *parser, const char *at, size_t len)
+{
+    Request *request = (Request *)parser->data;
+    request->SetPath(at, len);
+    return 0;
+}
+
+int Connection::OnHeaderField(http_parser *parser, const char *at, size_t len)
+{
+    Request *request = (Request *)parser->data;
+    request->AddHeaderField(at, len);
+    return 0;
+}
+
+int Connection::OnHeaderValue(http_parser *parser, const char *at, size_t len)
+{
+    Request *request = (Request *)parser->data;
+    request->AddHeaderValue(at, len);
+    return 0;
+}
+int Connection::OnBody(http_parser *parser, const char *at, size_t len)
+{
+    Request *request = (Request *)parser->data;
+    request->SetBody(at, len);
+    return 0;
+}
+
+int Connection::OnHeadersComplete(http_parser *parser)
+{
+    Request *request = (Request *)parser->data;
+    request->SetContentLength(parser->content_length);
+    request->SetMethod(parser->method);
+    return 0;
+}
+
+int Connection::OnMessageComplete(http_parser *parser)
+{
+    Request *request = (Request *)parser->data;
+    request->SetComplete(0);
```

Comment:
Then, when is the request `Complete`?

Reply:
parameter `complete` will be remove from request, because we only parse the buffer once and there will be only two results  :  error  occurs or parser successful, will add  error catching  later.